color/src/flags.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright 2024 the Color Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Types for tracking [`DynamicColor`](crate::DynamicColor) state.
+
+use crate::x11_colors;
+
+/// Flags indicating [`DynamicColor`](crate::DynamicColor) state.
+///
+/// The "missing" flags indicate whether a specific color component is missing (either the three
+/// color channels or the alpha channel).
+///
+/// The "named" flag represents whether the dynamic color was parsed from one of the named colors
+/// in [CSS Color Module Level 4 § 6.1][css-named-colors] or named color space functions in [CSS
+/// Color Module Level 4 § 4.1][css-named-color-spaces].
+///
+/// The latter is primarily useful for serializing to a CSS-compliant string format.
+///
+/// [css-named-colors]: https://www.w3.org/TR/css-color-4/#named-colors
+/// [css-named-color-spaces]: https://www.w3.org/TR/css-color-4/#color-syntax
+#[derive(Default, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct Flags {
+    /// A bitset of missing color components.
+    missing: Missing,
+
+    /// The named source a [`crate::DynamicColor`] was constructed from. Meanings:
+    /// - 0 - not constructed from a named source;
+    /// - 255 - constructed from a named color space function;
+    /// - otherwise - the 1-based index into [`crate::x11_colors::NAMES`].
+    name: u8,
+}
+
+// Ensure the amount of colors fits into the `Flags::name` packing.
+#[cfg(test)]
+const _: () = const {
+    if x11_colors::NAMES.len() > 253 {
+        panic!("There are more X11 color names than can be packed into Flags.");
+    }
+};
+
+/// Missing color components, extracted from [`Flags`]. Some bitwise operations are implemented on
+/// this type, making certain manipulations more ergonomic.
+#[derive(Default, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct Missing(u8);
+
+impl Flags {
+    /// Construct flags with the given missing components.
+    #[inline]
+    pub const fn from_missing(missing: Missing) -> Self {
+        Self { missing, name: 0 }
+    }
+
+    /// Set the missing components.
+    #[inline]
+    #[warn(
+        clippy::missing_const_for_fn,
+        reason = "can be made const with MSRV 1.83"
+    )]
+    pub fn set_missing(&mut self, missing: Missing) {
+        self.missing = missing;
+    }
+
+    /// Returns the missing components from the flags.
+    #[inline]
+    pub const fn missing(self) -> Missing {
+        self.missing
+    }
+
+    /// Set the flags to indicate the color was specified as one of the named colors. `name_ix` is
+    /// the index into [`crate::x11_colors::NAMES`].
+    pub(crate) fn set_named_color(&mut self, name_ix: usize) {
```

Comment:
This will never be called from user code, right? (I ask because otherwise `color_name` could panic in release mode)

Reply:
That's right, only the parser calls this.